src/test/java/com/google/devtools/build/lib/exec/CompactSpawnLogContextTest.java

```diff
@@ -160,6 +155,79 @@ public void testSymlinkAction() throws IOException, InterruptedException {
                 .build());
   }
 
+  @Test
+  public void testRunfilesTreeReusedForTool() throws Exception {
```

Comment:
How easy would it be to add a test case for the "multiple test attempts" scenario?

Reply:
We could create a `RunfilesTree` in the test that returns `true` for `isMappingCached()`, but that doesn't really test anything beyond us calling that function. Any kind of more realistic test would probably need to be an integration test that defines an actual test target, runs multiple test attempts and then asserts on the log. That would require a lot of setup.

Reply:
SGTM